src/AstGen.zig

```diff
@@ -1722,6 +1722,55 @@ fn structInitExpr(
         }
     }
 
+    // First: The first found field.
+    // Duplicates: A list of following duplicate field names.
+    const DupeNameRecord = struct { first: Ast.TokenIndex, duplicates: ArrayListUnmanaged(Ast.TokenIndex) };
+
+    var duplicate_names = std.AutoHashMap(u32, DupeNameRecord).init(astgen.arena);
+    try duplicate_names.ensureTotalCapacity(@intCast(struct_init.ast.fields.len));
+
+    // // Basic check to maintain O(N) best case.
+    var any_duplicate = false;
+
+    for (struct_init.ast.fields) |field| {
+        const name_token = tree.firstToken(field) - 2;
+        const name_index = try astgen.identAsString(name_token);
+
+        // Is the variable name already set?
+        const gop = try duplicate_names.getOrPut(name_index);
+
+        if (gop.found_existing) {
+            try gop.value_ptr.duplicates.append(astgen.arena, name_token);
+            any_duplicate = true;
+        } else {
+            gop.value_ptr.first = name_token;
+        }
+    }
+
+    if (any_duplicate) {
+        var it = duplicate_names.iterator();
```

Comment:
Ah, slight issue I just realised: hashmap iteration order is nondeterministic by design. You probably want to use an `ArrayHashMap` here to preserve insertion order.

...however, `ArrayHashMap` has multiple allocations involved, so deiniting will no longer reliably free from an arena. Okay, I'll give my final suggestion here in an overall review comment.